src/vault.cairo

```diff
@@ -0,0 +1,160 @@
+#[starknet::contract]
+mod Vault {
+    use core::num::traits::Zero;
+    use openzeppelin::access::ownable::OwnableComponent;
+    use openzeppelin::token::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+    use openzeppelin::upgrades::UpgradeableComponent;
+    use openzeppelin::upgrades::interface::IUpgradeable;
+    use spotnet::interfaces::IVault;
+    use spotnet::types::{TokenAmount};
+
+    use starknet::ContractAddress;
+    use starknet::storage::{
+        StoragePointerReadAccess, StoragePointerWriteAccess, StoragePathEntry, Map
+    };
+    use starknet::{ClassHash, get_caller_address, get_contract_address};
+
+    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);
+    component!(path: UpgradeableComponent, storage: upgradeable, event: UpgradeableEvent);
+
+    /// Ownable
+    #[abi(embed_v0)]
+    impl OwnableTwoStepMixinImpl =
+        OwnableComponent::OwnableTwoStepMixinImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+
+    /// Upgradeable
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        token: ContractAddress,
+        amounts: Map<ContractAddress, u256>,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+        LiquidityAdded: LiquidityAdded,
+        LiquidityWithdrawn: LiquidityWithdrawn,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct LiquidityAdded {
+        #[key]
+        user: ContractAddress,
+        #[key]
+        token: ContractAddress,
+        amount: u256,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct LiquidityWithdrawn {
+        #[key]
+        user: ContractAddress,
+        #[key]
+        token: ContractAddress,
+        amount: u256,
```

Comment:
Same

Reply:
I can't put the type inside Event because the `TokenAmount` lacks some impls @faurdent